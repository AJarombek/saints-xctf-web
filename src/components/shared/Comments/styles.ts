/**
 * JSS styles for the Comments component.
 * @author Andrew Jarombek
 * @since 8/5/2020
 */

import color from 'color';
import Colors, {FeelColors} from "../../../styles/colors";
import {FontMixins} from "../../../styles/mixins";

export default {
    comments: {
        display: 'flex',
        alignItems: 'center',
        flexDirection: 'column',
    },
    newCommentForm: {
        display: 'flex',
        alignItems: 'center',
        width: '100%',
    },
    newComment: {
        margin: '0 2%',
        padding: '6px 4px',
        height: '35px',
        backgroundColor: '#fdfdfd',
        border: ({feel}: {feel: number}) => `2px solid ${color(FeelColors[feel - 1]).darken(0.20).hex()}`,
        borderRadius: '5px',
        ...FontMixins.roboto(),
        fontSize: '14px',
        resize: 'none',

        '&:focus': {
            outline: 'none',
        }
    },
    focusNewComment: {
        width: 'calc(100% - 70px)',
    },
    blurNewComment: {
        width: '100%',
    },
    newCommentDisabled: {
        backgroundColor: '#eee',
    },
    addIcon: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',

        borderRadius: '50%',
        border: 'none',
        backgroundColor: Colors.spotPaletteBlue,
        boxShadow: '0 2px 4px 0 rgba(0, 0, 0, 0.1)',
        transition: 'box-shadow 0.4s ease, background-color 0.2s ease',
        color: 'white',
        cursor: 'pointer',
        height: 30,
        width: 30,

        '&:hover': {
            boxShadow: '0 2px 4px 0 rgba(0, 0, 0, 0.2)',
        },

        '&:active': {
            backgroundColor: color(Colors.spotPaletteBlue).darken(0.2).hex(),
        },

        '&:focus': {
            outline: 'none',
        },

        '& > p': {
            ...FontMixins.elegantIcons(),
            fontSize: 24,
            margin: 0,
            textAlign: 'center',
        },
    },
    addIconDisabled: {
        backgroundColor: '#ddd',

        '&:hover': {
            boxShadow: 'none',
        },
    },
    commentList: {
        display: 'flex',
        margin: '10px 0 0 0',
        width: '100%',
        flexDirection: 'column',
    },
    comment: {
        width: '100%',
        borderTop: '1px solid #AAA',
        paddingTop: '2px',
        paddingBottom: '8px',
    },
    commentHeader: {
        display: 'flex',
        alignItems: 'center',
        width: '100%',
    },
    titleLink: {
        ...FontMixins.robotoSlab(),
        color: '#333',
        textDecoration: 'underline',
        cursor: 'pointer'
    },
    date: {
        ...FontMixins.robotoBold(),
        fontSize: '14px',
        margin: '0 0 0 auto',
    },
    commentBody: {
        margin: '5px 0',
        display: 'inline-block',
        ...FontMixins.robotoBold(),
        fontSize: '14px',

        '& > a': {
            display: 'inline-block',
            ...FontMixins.robotoBold(),
            color: '#444',
            transition: 'color 0.4s ease',

            '&:hover': {
                color: color(Colors.spotPaletteBrown).darken(0.2).hex(),
            }
        }
    },
};
