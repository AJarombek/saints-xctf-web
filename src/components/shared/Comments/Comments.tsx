/**
 * Comments component which displays a list of comments and allows for a new comment to be created.
 * @author Andrew Jarombek
 * @since 8/5/2020
 */

import React, { useEffect, useRef, useState } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import { Comment, NewComments, RootState, User } from '../../../redux/types';
import classNames from 'classnames';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { parseTagsInText } from '../../../utils/logs';
import DefaultErrorPopup from '../DefaultErrorPopup/DefaultErrorPopup';
import { useDispatch, useSelector } from 'react-redux';
import { addComment, addCommentToFeed, postComment } from '../../../redux/modules/logs';
import { postNotification } from '../../../redux/modules/notifications';

interface Props {
  comments: Comment[];
  feel: number;
  logId: number;
  logUsername: string;
  user: User;
  inFeed: boolean;
  page?: number;
  filterBy?: string;
  bucket?: string;
  index?: number;
}

const useStyles = createUseStyles(styles);

const Comments: React.FunctionComponent<Props> = ({
  comments,
  feel,
  logId,
  logUsername,
  user,
  inFeed,
  page,
  filterBy,
  bucket,
  index
}) => {
  const classes = useStyles({ feel });

  const dispatch = useDispatch();
  const newComments: NewComments = useSelector((state: RootState) => state.logs.newComments);

  const [content, setContent] = useState('');
  const [showError, setShowError] = useState(false);
  const [prevErrorTime, setPrevErrorTime] = useState(0);
  const [isCreating, setIsCreating] = useState(false);

  const textAreaRef = useRef(null);

  useEffect(() => {
    if (newComments) {
      const newComment = Object.entries(newComments).filter(([commentLogId, _]) => +commentLogId === logId);

      if (newComment.length) {
        if (newComment[0][1].serverError && newComment[0][1].lastUpdated !== prevErrorTime) {
          setShowError(true);
          setPrevErrorTime(newComment[0][1].lastUpdated);
        }

        if (!newComment[0][1].isFetching) {
          setIsCreating(false);

          const content = textAreaRef.current.value;

          if (inFeed) {
            dispatch(
              addCommentToFeed(logId, content, user.username, user.first, user.last, filterBy, bucket, page, index)
            );
          } else {
            dispatch(addComment(logId, content, user.username, user.first, user.last));
          }

          textAreaRef.current.value = '';
        }
      }
    }
  }, [
    bucket,
    dispatch,
    filterBy,
    inFeed,
    index,
    logId,
    newComments,
    page,
    prevErrorTime,
    user.first,
    user.last,
    user.username
  ]);

  const onTextAreaKeyUp = (e: React.KeyboardEvent<HTMLTextAreaElement>): void => {
    e.currentTarget.style.height = '25px';
    e.currentTarget.style.height = `${e.currentTarget.scrollHeight + 4}px`;
  };

  const onTextAreaChange = (e: React.ChangeEvent<HTMLTextAreaElement>): void => {
    setContent(e.target.value);
  };

  const onCreateComment = async (content: string, user: User): Promise<void> => {
    if (content) {
      setIsCreating(true);
      const added = await dispatch(postComment(logId, user.username, user.first, user.last, content));

      if (added) {
        if (user.username !== logUsername) {
          dispatch(
            postNotification(
              logUsername,
              `${user.first} ${user.last} commented on your exercise log.`,
              `/log/view/${logId}`
            )
          );
        }

        const tagRegex = /@(?<username>[a-zA-Z0-9]+)/g;
        let matches = [];

        while ((matches = tagRegex.exec(content)) !== null) {
          const username = matches.groups.username;
          dispatch(
            postNotification(username, `${user.first} ${user.last} mentioned you in a comment.`, `/log/view/${logId}`)
          );
        }
      }
    }
  };

  return (
    <div className={classNames(classes.comments, 'comments')}>
      <div className={classes.newCommentForm}>
        <textarea
          className={classNames(
            classes.newComment,
            content ? classes.focusNewComment : classes.blurNewComment,
            isCreating && classes.newCommentDisabled
          )}
          maxLength={1000}
          placeholder="Comment"
          ref={textAreaRef}
          onChange={onTextAreaChange}
          onKeyUp={onTextAreaKeyUp}
          disabled={isCreating}
        />
        {!!content && (
          <button
            className={classNames('addIcon', classes.addIcon, isCreating && classes.addIconDisabled)}
            onClick={(): Promise<void> => onCreateComment(content, user)}
            disabled={isCreating}
          >
            <p>&#x4c;</p>
          </button>
        )}
      </div>
      {comments && (
        <div className={classNames('commentList', classes.commentList)}>
          {comments.map((comment) => (
            <div className={classNames('comment', classes.comment)} key={comment.comment_id}>
              <div className={classes.commentHeader}>
                <Link to={`/profile/${comment.username}`} className={classes.titleLink}>
                  {comment.first} {comment.last}
                </Link>
                <p className={classes.date}>{moment(comment.time).format('MMM. Do, YYYY h:mm A')}</p>
              </div>
              <div className={classes.commentBody}>{parseTagsInText(comment.content)}</div>
            </div>
          ))}
        </div>
      )}
      {showError && (
        <DefaultErrorPopup
          message="An unexpected error occurred while adding a comment"
          onClose={(): void => setShowError(false)}
        />
      )}
    </div>
  );
};

export default Comments;
