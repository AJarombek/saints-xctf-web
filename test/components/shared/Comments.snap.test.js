/**
 * Snapshot test for the Comments component.
 * @author Andrew Jarombek
 * @since 8/17/2020
 */

import React from 'react';
import Comments from '../../../src/components/shared/Comments';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <Comments
      comments={[
        {
          comment_id: 1,
          username: 'andy',
          first: 'Andy',
          last: 'Jarombek',
          log_id: 3,
          time: '2020-08-17 19:00:00',
          content: 'Test comment.'
        }
      ]}
      feel={6}
      postComment={() => {}}
      addComment={() => {}}
      newComments={{}}
      logId={2}
      user={{}}
      page={1}
      filterBy="all"
      bucket="all"
      index={1}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
