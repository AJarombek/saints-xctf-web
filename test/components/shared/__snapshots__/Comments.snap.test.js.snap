// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="comments-0-1-1 comments"
>
  <div
    className="newCommentForm-0-1-2"
  >
    <textarea
      className="newComment-0-1-3 newComment-d0-0-1-15 blurNewComment-0-1-5"
      disabled={false}
      maxLength={1000}
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Comment"
    />
  </div>
  <div
    className="commentList commentList-0-1-9"
  >
    <div
      className="comment comment-0-1-10"
    >
      <div
        className="commentHeader-0-1-11"
      >
        <a
          className="titleLink-0-1-12"
          href="/profile/andy"
          onClick={[Function]}
        >
          Andy
           
          Jarombek
        </a>
        <p
          className="date-0-1-13"
        >
          Aug. 17th, 2020 7:00:00 PM
        </p>
      </div>
      <div
        className="commentBody-0-1-14"
      >
        Test comment.
      </div>
    </div>
  </div>
</div>
`;
